src/components/controls/datepicker/DatePicker.js

```diff
@@ -0,0 +1,324 @@
+      new Date(startingYear, getMonth(todaysDate), getDate(todaysDate)) :
+      todaysDate;
+
+    this.state = {
+      descentAmount,
+      earliestPossibleDate,
+      latestPossibleDate,
+      selectionMode: determinedSelectionMode,
+      dateSelectionComplete: false,
+      suppressYearSelection: determinedSelectionMode === 'month',
+      suppressMonthSelection: determinedSelectionMode === 'day',
+      currentlySelectedDate,
+      currentlyDisplayedMonth: latestPossibleDate,
+      yearChunks,
+      currentYearChunkIndex: picker.getYearChunkIndex(yearChunks, startingYear || currentYear),
+      monthsToDisplayByYear: picker.splitMonthsByYears(years.reverse(), totalSelectableMonths, todaysDate),
+      dateWasPreselected,
+      ...selectedDateParts
+    };
+
+    this.yearSelected = this.yearSelected.bind(this);
+    this.monthSelected = this.monthSelected.bind(this);
+    this.daySelected = this.daySelected.bind(this);
+
+    this.setToPreviousYearChunk = this.setToPreviousYearChunk.bind(this);
+    this.setToNextYearChunk = this.setToNextYearChunk.bind(this);
+
+    this.setToPreviousYear = this.setToPreviousYear.bind(this);
+    this.setToNextYear = this.setToNextYear.bind(this);
+
+    this.setToPreviousMonth = this.setToPreviousMonth.bind(this);
+    this.setToNextMonth = this.setToNextMonth.bind(this);
+
+    this.setHeaderToYearChunkView = this.setHeaderToYearChunkView.bind(this);
+    this.setHeaderToMonthView = this.setHeaderToMonthView.bind(this);
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    return !isEqual(this.state.currentlySelectedDate, nextState.currentlySelectedDate);
+  }
+
+  componentDidUpdate() {
+    const { currentlySelectedDate, dateSelectionComplete } = this.state;
+    if (dateSelectionComplete) {
+      this.props.dateSelected(currentlySelectedDate);
+    }
+  }
+
+  setToPreviousYearChunk() {
+    this.setState(pickerStateTransitions.setToPreviousYearChunk);
+  }
+
+  setToNextYearChunk() {
+    this.setState(pickerStateTransitions.setToNextYearChunk);
+  }
+
+  setToPreviousYear() {
+    this.setState(pickerStateTransitions.setToPreviousYear);
+  }
+
+  setToNextYear() {
+    this.setState(pickerStateTransitions.setToNextYear);
+  }
+
+  setToPreviousMonth() {
+    this.setState(pickerStateTransitions.setToPreviousMonth);
+  }
+
+  setToNextMonth() {
+    this.setState(pickerStateTransitions.setToNextMonth);
+  }
+
+  setHeaderToYearChunkView() {
+    this.setState(pickerStateTransitions.setToYearSelection);
+  }
+
+  setHeaderToMonthView() {
+    this.setState({ selectionMode: 'month' });
+  }
+
+  getPickerGridValues() {
+    const {
+      currentlySelectedDate,
+      yearChunks,
+      currentYearChunkIndex,
+      selectionMode
+    } = this.state;
+
+    const currentYearChunk = yearChunks[currentYearChunkIndex];
+    const firstYearInChunk = first(currentYearChunk);
+    const lastYearInChunk = last(currentYearChunk);
+
+    if (selectionMode === 'month') {
```

Comment:
I thought about it and the reason I didn't is because there's a lot of dependencies connected throughout this function. I got a little lazy towards the end to be honest, so maybe it's worth another look.